perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuthzResolverImpl.java

```diff
@@ -87,6 +91,43 @@ public class AuthzResolverImpl implements AuthzResolverImplApi {
 
 	private static final Pattern columnNamesPattern = Pattern.compile("^[_0-9a-zA-Z]+$");
 
+
+	private static final RowMapper<Map<String, Map<Integer, Group>>> AUTHZ_COMPLEMENTARY_OBJECTS_AUTH_GROUPS_MAPPER = (rs, i) -> {
+		try {
+			Map<String, Map<Integer, Group>> perunBeansWithAuthorizedGroups = new HashMap<>();
+
+			// Iterate through all returned columns and try to extract PerunBean name from the labels
+			for (int j = rs.getMetaData().getColumnCount(); j > 0; j--) {
+				Matcher matcher = patternForExtractingPerunBean.matcher(rs.getMetaData().getColumnLabel(j).toLowerCase());
+				if (matcher.find()) {
+					String perunBeanName = matcher.group(1);
+					int complementaryObjectId = rs.getInt(j);
+
+					if (!rs.wasNull()) {
+						// We have to make first letters o words uppercase
+						String className = convertUnderScoreCaseToCamelCase(perunBeanName);
+
+						perunBeansWithAuthorizedGroups.computeIfAbsent(className, k -> new HashMap<>());
+
+						perunBeansWithAuthorizedGroups.get(className).put(
+							complementaryObjectId,
+							new Group(
+								rs.getInt("authz_authorized_group_id"),
+								rs.getString("groups_name"),
+								rs.getString("groups_description"),
+								rs.getInt("groups_vo_id")
+							));
+					}
+				}
+			}
+
+			return perunBeansWithAuthorizedGroups;
```

Comment:
This code is quite difficult to read, so I might be wrong but I think you need to return a structure like this: 
Pair<RoleName, Map<AuthGroupId, Pair<BeanName, Set\<ObjectId>>>>
Otherwise you completely lose the information about which role is associated with which authorized group / object, no?

Reply:
I admit, code is not the best one to read and yes, if you used only this method, you would lose information about which role is connected to which object.
However, when you get all roles for user, they contain an id of complementary object connected to it, so it can be used to obtain specific groups from map returned by `getRoleComplementaryObjectsWithAuthorizedGroups` method.

Reply:
Right, but can't a user have multiple roles in the same complementary object through different authorized groups? How do you then tell which group provides which role?

Reply:
I guess there could be a way to get this information through a combination of the methods, but that seems like extra logic for no reason.